File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;

        while (head != null && head.next != null) {
            slow = slow.next;
            head = head.next.next;

            if (slow == head) {
                return true;
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        if (lists.isEmpty()) {
            return null
        }
        var step = 1
        while (`step` < lists.size) {
            for (i in 0 until (lists.size - `step`) step `step`*2) {
                lists[i] = merge(lists[i], lists[i + `step`])
            }
            `step` *= 2
        }

        return lists[0]
    }

    private fun merge(l: ListNode?, r: ListNode?): ListNode? {
        var left: ListNode? = l
        var right: ListNode? = r
        val dummy = ListNode(0)
        var current: ListNode? = dummy
        while (left != null && right != null) {
            if (left.`val` <= right.`val`) {
                current?.next = left
                left = left.next
                current = current?.next
            } else {
                current?.next = right
                right = right.next
                current = current?.next
            }
        }

        current?.next = if (left == null) right else left
        return dummy.next
    }
}
